### ISSUE TITLE
Uncaught reference error

### ISSUE THREAD

**Issue by ntenisOT** [pagespeed/ngx_pagespeed#1219]
Hello and the for this great tool.

We have a setup with multiple servers behind Load balancing IP for our site ipvdx.com.

Occasionally we get:

    home:2218 Uncaught ReferenceError: mod_pagespeed_mEIYH0Ztx8 is not defined
    home:2219 Uncaught ReferenceError: mod_pagespeed_xB7sV__Veq is not defined
    home:2220 Uncaught ReferenceError: mod_pagespeed__awIdTW9XL is not defined
    home:2221 Uncaught ReferenceError: mod_pagespeed_DHZ4sFp5BJ is not defined
    home:2222 Uncaught ReferenceError: mod_pagespeed_J3XEBfdPXx is not defined
    home:2223 Uncaught ReferenceError: mod_pagespeed_ZUHQ_VWckV is not defined
    home:2224 Uncaught ReferenceError: mod_pagespeed__PnV7XRqUI is not defined
    home:2225 Uncaught ReferenceError: mod_pagespeed_nCHUuArfKA is not defined
    home:2226 Uncaught ReferenceError: mod_pagespeed_G8NvsT0wGU is not defined

We thought that this error might occur if for example the one file has been created on one server and we try to access it from the second server but I changed the setup and used GlusterFS to store the pagespeed cache but that does not solve the problem. In order to get rid of the error I just have to delete all cache and pagespeed will rebuilt it but it is annoying. 

Is there a way to get rid of that error or is it a known bug?

**Comment by oschaaf** [comment]
A similar issue was reported for mod_pagespeed:
https://github.com/pagespeed/mod_pagespeed/issues/1316#issuecomment-223656579

@jmarantz suggests a workaround here: https://github.com/pagespeed/mod_pagespeed/issues/1316#issuecomment-223656579

Right now I'm not able to offer anything more informative because I find it hard to debug/assess the problem right now (the problem doesn't seem to be happening at ipvdx.com at the moment).

**Comment by jmarantz** [comment]
The problem looks related to combine_javascript and there have been a few other complaints about this symptom.  We need to look into it.  If you can provide any more information about how & when this occurs, it would help a lot.

It is not related to having multiple servers.  PageSpeed is designed to be allow for separate caches on each server, and in fact I would strongly recommend *not* using a shared file-system to share the PageSpeed cache.  That can lead to very poor performance.   You can configure PageSpeed to use memcached however, to get the benefits of a shared cache.

**Comment by ntenisOT** [comment]
@jmarantz If you have two servers and you don't use any stickiness ( session, cookie, ip, etc ) what guarantees you that every request will be served by the same server? Nothing. Is that a problem for pagespeed?

**Comment by jmarantz** [comment]
That is not a problem for PageSpeed.  If server1 rewrites HTML to create a
reference to X.pagespeed.filter.hash.extension, and the request for that
resource is sent to server2, everything that's needed to decode and
reconstruction is in the encoding "X".  For example, combined css files
will look like a.css+b.css.pagespeed.cc.HASH.css.  That syntax is
decodeable and reconstructable with a clean cache.

Of course performance will be better once server2's cache is filled, and
that's why we have memcached as an option, so that server1 can write into a
cache visible by server2.

-Josh

**Comment by morlovich** [comment]
It could be a problem if the server aren't quite in sync.. Hmm. Is there any URL remapping going on between the load-balancer and the two servers?

**Comment by ntenisOT** [comment]
@morlovich I will make sure that whenever the problem occurs I will give you the link and everything. I will also point the domains directly to individual servers for you to check. 

@jmarantz I tried to do what you describe using Glusterfs ( Ok it is not as fast as memcache but if you refresh the page after 1 minute the files should be there ) but it didn't work. So it should be another problem. Additionally, we use stickiness using IP and Cookies aswell.

**Comment by jmarantz** [comment]
Using a networked file system isn't just a system-speed issue.  Our
file-cache reading code is not threaded, and while waiting for networked
resource will block all other activities on our 'rewrite' thread, which is
a bounded resources.

In contrast, our memcached interface operates in its own thread, and has
health monitoring and timeouts to ensure that a frozen memcached doesn't
freeze your web server.


We definitely need to get to the bottom of the combine_javascript reference
errors though.

-Josh

**Comment by ntenisOT** [comment]
@jmarantz  Thanks a lot for the clarification. If an error occurs again I will provide you all information and try the memcached in-memory solution as well.

**Comment by morlovich** [comment]
I should make my question a bit clearer: I think things may get confused if the two servers think the URLs for the component JS files are different.

**Comment by ntenisOT** [comment]
@morlovich  We use stickiness for our Load Balancer using Source IP (https://www.ovh.ie/solutions/ip-load-balancing/) and we had the same issues as well. That's why I tried network storage as a fallback solution and that's why I asked here because we are unable to solve the problem since we are unable to identify where the problem is.

**Comment by MiguelQueiroz** [comment]
I have this problem, and i dont use load balacing.. only one server. It annoying because it fills my "Error reporting tool" with tons of javascript errors, that will "overflow" the graph, with real issues...

**Comment by oschaaf** [comment]
@MiguelQueiroz having a live display of the bug happening would be really helpful. Could you post an url here (or email me at oschaaf@we-amp.com if you are uncomfortable with sharing the url here) when the problem happens?

**Comment by MiguelQueiroz** [comment]
Yes sure.. You go to our company website http://zappysoftware.com/ then click on pricing. But open the dev tools.. before clicking pricing. And you will see that, the page wont load via ajax, there an error on the console like that one reported. You can navigate on other tabs and see that they load, but they usually display that same error also. And im running the latest version. I believe i will need to remove pagespeed...

**Comment by oschaaf** [comment]
@MiguelQueiroz thanks!

Question:  Did you, at some point, directly delete contents from ngx_pagespeed's filesystem cache folder?

Some observations so far:

The page chokes on `mod_pagespeed_alawqg0Ntp` not being defined.

When I request `/pricing` with `?PageSpeedFilters=+debug` the error is not present.

With debug enabled, mod_pagespeed_alawqg0Ntp gets defined at the start of: 
```
/wp-includes,_js,_jquery,_ui,_jquery.ui.core.min.js,qver==1.10.4+wp-includes,_js,_jquery,_ui,_jquery.ui.widget.min.js,qver==1.10.4+wp-includes,_js,_jquery,_ui,_jquery.ui.tabs.min.js,qver==1.10.4+wp-content,_plugins,_js_composer,_assets,_lib,_jquery-ui-tabs-rotate,_jquery-ui-tabs-rotate.js,qver==4.2.3.pagespeed.jc.LDla14CUPD.js`
```

That combined url is not written in the non-debug version which gave the error.
It gives response:

```
var mod_pagespeed_alawqg0Ntp = "(function(e,t){function i(t,i){var a,n,r,o=t.
```

The original js url in question to which I think `mod_pagespeed_alawqg0Ntp` corresponds is: /wp-content/themes/bridge/js/plugins.js?ver=3.9.2

This file indeed is not combined, but shows up like this in the html, mostly untouched:
```
<script type='text/javascript' src='http://zappysoftware.com/wp-content/themes/bridge/js/plugins.js?ver=3.9.2'></script>
```

On a sidenote, another observation is that now that I have requested the page with debug enabled, the problem seems to have gone away for /pricing.

**Comment by oschaaf** [comment]
Furthermore, on `/wellness-software` in `onload` I see another script error, all by its own:
```
VM497 wellness-software:1 Uncaught ReferenceError: pagespeed is not defined
```

Offending code:
```
(function(event){pagespeed.CriticalImages.checkImageForCriticality(this);
})
```

Dumped the html here for inspection:
https://gist.github.com/oschaaf/edcec9575d297411b4713813d85ac1a1

**Comment by MiguelQueiroz** [comment]
@oschaaf  Right, pagespeed too. These errors does not happen always. Dont know why it sometimes does...

**Comment by oschaaf** [comment]
@MiguelQueiroz Did you manually delete files and/or directories from ngx_pagespeed's cache folder?

**Comment by MiguelQueiroz** [comment]
no

**Comment by oschaaf** [comment]
@MiguelQueiroz Thanks. I noticed that the problem only happens when navigating on the site, and not when switching pages via the browser.  

It looks like your website is using XMLHttpRequest + javascript to load in pages upon navigating via the menu (so adding debug did not change anything - but the browser navigation following from navigation the querystring did). 

They way your website performs its navigation is probably is not going to work well with some of mod_pagespeed's filters, and it looks like combining is one of them..

**Comment by MiguelQueiroz** [comment]
Yes ajax loading, which is very common in wordpress websites.

**Comment by oschaaf** [comment]
@MiguelQueiroz 
Thanks. I think OptimizeForBandwidth may work well for you as a starting point: https://developers.google.com/speed/pagespeed/module/optimize-for-bandwidth

**Comment by MiguelQueiroz** [comment]
@oschaaf  is possible to put the js between try and catch tags? to prevent from creating the "undefined" error on logs?

**Comment by jmarantz** [comment]
I don't think hiding the log messages with a try/catch is the right strategy; we need to fix them, because the 'reference error' messages are telling us that the combined JavaScript is not loading.

We had been trying to figure out what causes this bug, and I guess it's pretty simple: your combined JS files are failing to load because when you are in an Ajax load, the base URL of the page the browser sees has a different path from the base URL of the page that ngx_pagespeed rewrites on the server.  It that's the case, then there should also be console log messages in the browser telling us that the combined JS file failed to load (404).  Is that happening?

Typically Ajax-loads will have a request header of X-Requested-With :XMLHttpRequest .  We should double-check that when this occurs, we doing the right thing with respect to absolute/relative rewritten URLs.

**Comment by MiguelQueiroz** [comment]
In my case, no there are no 404 errors.  Maybe because, my "theme" will just load "HTML" content, and strips every <script> tags. It does so, because the ajax loads only content. And believes that every script is already on the parent scope loaded, so you dont get duplicate script loads... So thats why, it wont complain about 404 scripts missing, because, it actually wont try to load them. However, it will try to call the pagespeed "uncompress" function, and wont find the defined script.
